### PR TITLE
Add support for spending keys to the basic key store

### PR DESCRIPTION
**Pull request by str4d**
Prerequisite for #1197 and #1199.

### CODE REVIEW

**Comment by ebfull** [comment]
NACK until I review :)

**Comment by ebfull** [comment]
Good work! This definitely needs to combine `AddSpendingKey` and `AddSpendingKeyPaymentAddress` together.

File: src/keystore.h [hunk 1]
```diff
@@ -103,6 +115,43 @@ class CBasicKeyStore : public CKeyStore
...
+            SpendingKeyMap::const_iterator mi = mapSpendingKeys.find(address);
+            if (mi != mapSpendingKeys.end())
+            {
+                keyOut = mi->second;
+                return true;
+            }
+        }
+        return false;
+    }
+    void GetPaymentAddresses(std::set<libzcash::PaymentAddress> &setAddress) const
```
Comment (ebfull) [inline src/keystore.h]: This API appears to only ever be used so far in your other PR on note tracking, but I think we should be creating `ZCNoteDecryption` objects for each spending key and caching those for use in checking input notes, and keeping them consistent with the spending keys currently added to or imported to the wallet.

Comment (ebfull) [inline src/keystore.h]: Actually it's used in the `z_listaddresses` RPC as well, so that's fine.

Comment (str4d) [inline src/keystore.h]: I'm now caching `ZCNoteDecryption` objects in #1233.

**Comment by str4d** [comment]
I've looked over how `AddKeyPubKey()` is used for transparent addresses. Aside from its obvious use in `AddKey()`, it is used in the following locations:

- `wallet/rpcdump.cpp` - inside the RPC commands `importprivkey` and `importwallet`. Both of these call `CPubKey pubkey = key.GetPubKey()` followed by `assert(key.VerifyPubKey(pubkey))` before calling `pwalletMain->AddKeyPubKey(key, pubkey)`.
- `wallet/crypter.cpp` - inside the overridden `AddKeyPubKey()`. If the wallet is unencrypted, it calls through to the base implementation. If the wallet is encrypted, they encrypt the key and then call `AddCryptedKey(pubkey, vchCryptedSecret)`.

The only reason I can imagine for passing the pubkey through the API is for efficiency. In `rpcdump.cpp` in particular, `pubkey` is used to get the `CKeyID`, which is used to set some metadata before `AddKeyPubKey()` is called. By passing in `pubkey`, it is only generated once (rather than generating it both outside and inside the API call).

**Comment by str4d** [comment]
In the case of `crypter.cpp` there is no efficiency concern. Either the base `AddKey` could be called inside which the public key is generated, or the public key would need to be generated before encrypting the key and storing. These are separate branches, and so via either branch `pubkey` is only generated once.

**Comment by str4d** [comment]
So the decision here is between efficiency and having an API that can't be misused. I don't think that `importwallet` or the Zcash analogue of `importprivkey` will be used often enough to justify the efficiency gain, and so I agree with @ebfull that the methods should be merged.

**Comment by str4d** [comment]
@ebfull changes made. I'll rebase #1233 once this has been ACKed and merged.

**Comment by daira** [comment]
ACK

**Comment by ebfull** [comment]
ACK

**Comment by str4d** [comment]
Still needs an ACK+cov :smiley:

**Comment by bitcartel** [comment]
Can we have more fine grained locking e.g. cs_spending_lock? The new methods which mutate the spending key map don't touch existing members protected by cs_lock

Not a critical issue but might be worth considering if there is enough time. 

ACK

File: src/keystore.h [hunk 2]
```diff
@@ -103,6 +115,43 @@ class CBasicKeyStore : public CKeyStore
...
+        {
+            LOCK(cs_SpendingKeyStore);
+            SpendingKeyMap::const_iterator mi = mapSpendingKeys.find(address);
+            if (mi != mapSpendingKeys.end())
+            {
+                skOut = mi->second;
+                return true;
+            }
+        }
+        return false;
```
Comment (ebfull) [inline src/keystore.h]: There is no test coverage for failing to return a spending key.

**Comment by ebfull** [comment]
ACK+cov modulo that nit.

**Comment by str4d** [comment]
Done!

@zkbot +r

**Comment by str4d** [comment]
Oops!

@zkbot r+

**Pull request merged**